internal/uvm/scsi/manager.go

```diff
@@ -0,0 +1,303 @@
+package scsi
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/protocol/guestresource"
+	"github.com/Microsoft/hcsshim/internal/verity"
+	"github.com/Microsoft/hcsshim/internal/wclayer"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	// ErrNoAvailableLocation indicates that a new SCSI attachment failed because
+	// no new slots were available.
+	ErrNoAvailableLocation = errors.New("no available location")
+	// ErrNotInitialized is returned when a method is invoked on a nil [Manager].
+	ErrNotInitialized = errors.New("SCSI manager not initialized")
+	// ErrAlreadyReleased is returned when [Mount.Release] is called on a Mount
+	// that had already been released.
+	ErrAlreadyReleased = errors.New("mount was already released")
+)
+
+// Manager is the primary entrypoint for managing SCSI devices on a VM.
+// It tracks the state of what devices have been attached to the VM, and
+// mounted inside the guest OS.
+type Manager struct {
+	attachManager *attachManager
+	mountManager  *mountManager
+}
+
+// Slot represents a single SCSI slot, consisting of a controller and LUN.
+type Slot struct {
+	Controller uint
+	LUN        uint
+}
+
+// NewManager creates a new Manager using the provided backend [Attacher],
+// [Mounter], and [Unplugger], as well as other configuration parameters.
+//
+// guestMountFmt is the format string to use for mounts of SCSI devices in
+// the guest OS. It should have a single %d format parameter.
+//
+// reservedSlots indicates which SCSI slots to treat as already used. They
+// will not be handed out again by the Manager.
+func NewManager(
+	attacher Attacher,
+	mounter Mounter,
+	unplugger Unplugger,
```

Comment:
if we are taking these directly from users, we probably should validate that they aren't nil.
and that they arent using a `bridgeUnplugger` with an `hcsMounter`

Reply:
Happy to add a `!= nil` validation.

I don't think it would be right to check them against specific implementations here, because we don't want the scsi code to care what backend implementations you pass it. It's entirely up to the client to pass correct backend values. In practice this should be an issue, because this is internal-only code, so we can expect the client to be knowledgeable about the correct way to use it. Managers are also constructed rarely (only in two places in our product code), so it's easy to validate with a spot check that we are doing it correctly.

Reply:
Im not understanding why we expose the low level unpluggers, since it clients shouldnt need to worry about specifics, simply that they want to use SCSI over GCS or a bridge, and I dont expect we would add any `unplugger`/`mounter`/`attacher` implementations outside of the package
theres isnt a reason for the client to be knowledgeable of how the SCSI internals work, right?

Reply:
How about making `scsi.NewManager` take a `hostBackend` and a `guestBackend`, which can be constructed with e.g. `scsi.NewHCSHostBackend`, `scsi.NewBridgeGuestBackend`? That way, we can keep the desired decoupling of host/guest operations, but it's still easier for the client to get the needed objects and pass them in?

Internally, `scsi.hcsHostBackend` would be a struct that implements `attacher`, and `scsi.bridgeGuestBackend` would be a struct that implements `mounter` and `unplugger`.

Reply:
I implemented this, and think it is an improvement. Hopefully this addresses the concerns over clarity of client interface.